Prevent unnecessary execution of the `displayValue` callback in the `ComboboxInput` component

RobinMalfait opened the pull request:
This PR fixes an issue where the `displayValue` on the `ComboboxInput` component is executed on every render. This PR fixes that by only executing the callback if either the `value` or the `displayValue` itself changes.

This is especially useful if you have a `displayValue` that is a function that does some heavy computation or that contains undeterministic code. E.g.: `Math.random()`, `Date.now()`, …

Fixes: #3044

thecrypticace commented:
We used to useMemo for `displayValue` but changed it. The big difference here is the dependency array though. We'll want to make sure this doesn't break: https://github.com/tailwindlabs/headlessui/pull/2090

RobinMalfait commented:
> We used to useMemo for `displayValue` but changed it. The big difference here is the dependency array though. We'll want to make sure this doesn't break: #2090 (this is the PR where we changed it to the current impl)

Aha good catch! But yep just updated the `useMemo` dependencies array where I removed the `displayValue` and that does indeed break the previous issue. Adding it though, makes the test pass.

When running: `clear; git diff; npm run test combobox; git reset --hard; npm run test combobox` you can see that the `displayValue` as a dependency is important.

<img width="1817" alt="image" src="https://github.com/tailwindlabs/headlessui/assets/1834413/c5b1b122-0d09-4df1-835d-a062c78a73cd">
